### PR TITLE
[no-commonjs] add allowConditionalRequire option

### LINKED ISSUE #1437: [no-commonjs] does not catch conditional require
The example here: https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-commonjs.md#allow-require

cannot be reported even if `allowRequire: false`, because of this line:

https://github.com/benmosher/eslint-plugin-import/blob/1edbbd03ec636469328ad59da922ed7edc8cd36d/src/rules/no-commonjs.js#L90

The context scope is `block` in this case.

### PR DESCRIPTION
**Pull request by Pessimistress**
Fixes #1437

Currently:

1. `if (typeof window !== "undefined") require("x")` is reported by default
1. `if (typeof window !== "undefined") { require("x") }` is never reported (scope)
1. `typeof window !== "undefined" && require("x")` is never reported (conditional require)

After this PR, all cases will be treated the same (as conditional requires). A follow up PR will add an option for users to configure whether conditional requires should be reported.

### CODE REVIEW

File: src/rules/no-commonjs.js [hunk 1]
```diff
@@ -25,6 +25,11 @@ function allowRequire(node, options) {
   return options.allowRequire
 }
 
+function validateScope(scope) {
+  if (scope.variableScope.type === 'module') return true
```
Comment (Pessimistress) [inline src/rules/no-commonjs.js]: I think this might break the current behavior with try... catch. I'm AFK right now, will add a test case later.

**Comment by Pessimistress** [comment]
I have no idea how to add the option without making these changes... Do you mind if I just add it in the same PR?

**Comment by Pessimistress** [comment]
> those who are relying on these warnings have an existing solution to enable them

I highly doubt if anyone is relying on these warnings. The only breaking change is that `if (typeof window !== "undefined") require("x")` is no longer reported. Considering `if (typeof window !== "undefined") { require("x") }` was never reported, I consider the existing behavior already broken.

**Comment by ljharb** [comment]
@Pessimistress adding the option in this PR sounds fine to me.

File: src/rules/no-commonjs.js [hunk 2]
```diff
@@ -25,6 +25,27 @@ function allowRequire(node, options) {
   return options.allowRequire
 }
 
+function allowConditionalRequire(node, options) {
+  return options.allowConditionalRequire === undefined ? true : options.allowConditionalRequire
```
Comment (ljharb) [inline src/rules/no-commonjs.js]:
```suggestion
  return options.allowConditionalRequire !== false
```

File: src/rules/no-commonjs.js [hunk 3]
```diff
@@ -25,6 +25,27 @@ function allowRequire(node, options) {
...
+}
+
+function validateScope(scope) {
+  if (scope.variableScope.type === 'module') return true
+  return false
```
Comment (ljharb) [inline src/rules/no-commonjs.js]:
```suggestion
  return scope.variableScope.type === 'module'
```

File: src/rules/no-commonjs.js [hunk 4]
```diff
@@ -35,6 +56,7 @@ const schemaObject = {
   properties: {
     allowPrimitiveModules: { 'type': 'boolean' },
     allowRequire: { 'type': 'boolean' },
+    allowConditionalRequire: { 'type': 'boolean' },
```
Comment (ljharb) [inline src/rules/no-commonjs.js]:
```suggestion
    allowConditionalRequire: { 'type': 'boolean', default: true },
```
(altho for back compat we can't rely on the default getting applied)

File: src/rules/no-commonjs.js [hunk 5]
```diff
@@ -25,6 +25,27 @@ function allowRequire(node, options) {
   return options.allowRequire
 }
 
+function allowConditionalRequire(node, options) {
+  return options.allowConditionalRequire !== false
+}
+
+function validateScope(scope) {
+  if (scope.variableScope.type === 'module') return true
```
Comment (ljharb) [inline src/rules/no-commonjs.js]:
```suggestion
```
now that the other change is done :-)

**Pull request merged**